include/oneapi/dpl/internal/scan_by_segment_impl.h

```diff
@@ -0,0 +1,372 @@
+                        else
+                        {
+                            __partials_acc[__group_id] = __accumulator;
+                        }
+                    }
+                });
+        });
+
+        // 2. Apply work group carry outs, calculate output indices, and load results into correct indices.
+        __exec.queue()
+            .submit([&](sycl::handler& __cgh) {
+                oneapi::dpl::__ranges::__require_access(__cgh, __keys, __out_values);
+
+                auto __partials_acc = __partials.template get_access<sycl::access_mode::read>(__cgh);
+                auto __seg_ends_acc = __seg_ends.template get_access<sycl::access_mode::read>(__cgh);
+
+                __cgh.depends_on(__wg_scan);
+
+                __dpl_sycl::__local_accessor<__val_type> __loc_partials_acc(__wgroup_size, __cgh);
+
+                __dpl_sycl::__local_accessor<__flag_type> __loc_seg_ends_acc(__wgroup_size, __cgh);
+#if _ONEDPL_COMPILE_KERNEL && _ONEDPL_KERNEL_BUNDLE_PRESENT
+                __cgh.use_kernel_bundle(__kernel2.get_kernel_bundle());
+#endif
+                __cgh.parallel_for<_SegScanKernel2>(
+#if _ONEDPL_COMPILE_KERNEL && !_ONEDPL_KERNEL_BUNDLE_PRESENT
+                    __kernel2,
+#endif
+                    sycl::nd_range<1>{__n_groups * __wgroup_size, __wgroup_size}, [=](sycl::nd_item<1> __item) {
+                        auto __group = __item.get_group();
+                        ::std::size_t __group_id = __group.get_group_id(0);
+                        ::std::size_t __global_id = __item.get_global_id();
+                        ::std::size_t __local_id = __item.get_local_id();
+                        ::std::size_t __start = __global_id * __vals_per_item;
+                        ::std::size_t __end = __dpl_sycl::__minimum<decltype(__n)>{}(__start + __vals_per_item, __n);
+
+                        ::std::size_t __wg_agg_idx = __group_id - 1;
+                        __val_type __agg_collector = __identity;
+
+                        //TODO:  just launch with one fewer group and adjust indexing since group zero can skip phase
+                        if (__group_id != 0)
+                        {
+                            // 2a. Calculate the work group's carry-in value.
+                            bool __ag_exists = false;
+                            if (__start < __n)
+                            {
+                                __ag_exists = true;
+                            }
+                            // local reductions followed by a downsweep
+                            // TODO: Generalize this value
+                            constexpr int64_t __vals_to_explore = 16;
+                            __flag_type __last_it = false;
+                            __loc_seg_ends_acc[__local_id] = false;
+                            __loc_partials_acc[__local_id] = __identity;
+
+                            for (int64_t __i = __wg_agg_idx - __vals_to_explore * __local_id;;
```

Comment:
The double `;;` is actually because the for loop test condition is omitted and the formatting makes it confusing. The for loop continues until a segment end has been identified and then breaks. I will look for a better way to structure this as I agree that it looks confusing.

The maximum value of `__i` is essentially the total number of work groups in the system. If we process `4` elements per work group and say we have a work group size of `1024`, then the total number of work groups hits `32,767` at an input size of `134,213,632`, and we would have an overflow with anything greater. I think it is reasonable to use `int32_t` instead, however.

Reply:
I have reformatted this for loop to be more clear: the break condition is now checked in the test condition instead of the end of the loop. 

`__i` is now `::std::uint32_t` which is sufficiently large for all possible number of work groups.